### PR TITLE
Persist addl claims+tokens w/ext providers

### LINKED ISSUE #3744: Persisting additional claims and tokens from external providers
When signing up for your application with an external provider (e.g. Facebook), the developer may want to persist additional data about the user, as well as their access and refresh tokens. 
@HaoK @blowdart @danroth27 

There are at least three parts to this process.
A. Extracting additional information from the provider and storing it as claims. The information is provider specific, but there is a generic pattern. In 1.x the developer hooks into an event on the auth middeware, reads the user info blob, and creates new claims for the temp cookie. For 2.0 the developer no longer needs to hook into the events to do this manually, there is a new ClaimActions list on the options that lets them add and remove mappings. See OAuthOptions.ClaimActions, OpenIdConnect.ClaimActions and TwitterOptions.ClaimActions.
B. Persisting access tokens in the temporary cookie. See RemoteAuthenticationOptions.SaveTokens.
C. When you get to the account controller, you need to copy this information into the user database on app sign-up, or optionally update it per sign-in.

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #3744 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/security/authentication/social/additional-claims?view=aspnetcore-2.0&branch=pr-en-us-8310)

* I made several *uneducated guesses* here. Let me know where I went off the rails. 🛠
* I haven't validated the non-Google scopes. If they aren't known by our folks, should we trust the sources I found that provided the values, or should I configure providers and confirm they work? It's time-consuming to check them, and we have a massive amount of work to do 🏃😅. However, I'll validate them myself if we need it done.
* Do you want a 1.x sample, too? If so, it would be best if it's produced for me or if someone points me to an existing 1.x identity sample where I can then attempt to apply Chris's notes: *"The developer hooks into an event on the auth middeware, reads the user info blob, and creates new claims for the temp cookie."*

### CODE REVIEW

**Comment by guardrex** [comment]
@HaoK Are you OOF right now? This can wait ... I can push the issue's completion back a sprint if we need to.

**Comment by HaoK** [comment]
No sorry, I just lose track of mentions easily these days, I'll take a look

**Comment by HaoK** [comment]
Looks good to me, but I'm probably a bit too familiar with this flow since its based off the sample I wrote :)  Adding @blowdart and @Tratcher for a pair of fresh eyes in case they see anything fishy.

File: aspnetcore/security/authentication/social/additional-claims.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+### Set the client ID and client secret
+
+The OAuth authentication provider establishes a trust relationship with an app using a client ID and client secret. Client ID and client secret values are created for the app by the external authentication provider when the app is registered with the provider. Each external provider that the app uses must be configured independently with the provider's client ID and client secret. For more information, see the external authentication provider topics that apply to your scenario:
+
+* [Facebook authentication](xref:security/authentication/facebook-logins)
+* [Google authentication](xref:security/authentication/google-logins)
+* [Microsoft authentication](xref:security/authentication/microsoft-logins)
+* [Twitter authentication](xref:security/authentication/twitter-logins)
+* [Other authentication providers](xref:security/authentication/otherlogins)
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/security/authentication/social/additional-claims.md]: AFAICT the only spot to link it is ...

https://docs.microsoft.com/en-us/aspnet/core/migration/1x-to-2x/identity-2x?view=aspnetcore-2.1#openid-connect-oidc-authentication

... so I'll do that.

File: aspnetcore/security/authentication/social/additional-claims/samples/2.x/AdditionalClaimsSample/Startup.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                options.Scope.Add("https://www.googleapis.com/auth/plus.login");
+                options.ClaimActions.MapJsonKey(ClaimTypes.Gender, "gender");
+                options.SaveTokens = true;
+                options.Events.OnCreatingTicket = ctx =>
+                {
+                    List<AuthenticationToken> tokens = ctx.Properties.GetTokens() 
+                        as List<AuthenticationToken>;
+                    tokens.Add(new AuthenticationToken()
+                    {
+                        Name = "TicketCreated",
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/security/authentication/social/additional-claims/samples/2.x/AdditionalClaimsSample/Startup.cs]: @HaoK ??

File: aspnetcore/security/authentication/social/additional-claims.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tratcher) [reply aspnetcore/security/authentication/social/additional-claims.md]: Bleh, that's worse than providing no link at all. How about https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2?

File: aspnetcore/security/authentication/social/additional-claims/samples/2.x/AdditionalClaimsSample/Startup.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (HaoK) [reply aspnetcore/security/authentication/social/additional-claims/samples/2.x/AdditionalClaimsSample/Startup.cs]: That is just demonstrating how you can add custom tokens which get stored as part of SaveTokens so they show up on the identity side as a token

**Comment by guardrex** [comment]
@Tratcher How's that look?

File: aspnetcore/security/authentication/social/additional-claims.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+* Obtain the user's gender from Google and store a gender claim with the value.
+* Store the Google access token in the user's `AuthenticationProperties`.
+
+To use the sample app:
+
+1. Register the app and obtain a valid client ID and client secret for Google authentication. For more information, see <xref:security/authentication/google-logins>.
+1. Provide the client ID and client secret to the app in the <xref:Microsoft.AspNetCore.Authentication.Google.GoogleOptions> of `Startup.ConfigureServices`.
+1. Run the app and request the My Claims page. When the user isn't signed in, the app redirects to Google. Sign in with Google. Google redirects the user back to the app (`/Home/MyClaims`).
+1. The user is authenticated, and the My Claims page is loaded. The gender claim is present under **User Claims** with the value obtained from Google. The access token appears in the **Authentication Properties**.
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/security/authentication/social/additional-claims.md]: I can roll this into the prior step. It all flows from the same user action (i.e., they run the app and request the page).

**Pull request merged**